plugins/single_plugins/idle.cpp

```diff
@@ -4,69 +4,250 @@ extern "C"
 }
 
 #include "singleton-plugin.hpp"
+#include "render-manager.hpp"
 #include "output.hpp"
 #include "core.hpp"
 #include "config.hpp"
 #include "output-layout.hpp"
+#include "../cube/cube-control-signal.hpp"
+
+#include <animation.hpp>
+
+#define ZOOM_BASE 1.0
 
 class wayfire_idle
 {
+    wf::effect_hook_t render_hook;
+    double rotation = 0.0;
+    wf_duration duration;
+    wf_transition rot_transition, zoom_transition;
+    bool screensaver_stopping = false;
     bool idle_enabled = true;
-    wlr_idle_timeout *timeout = NULL;
-    wf::wl_listener_wrapper on_idle, on_resume;
+    bool screensaver_enabled = false;
```

Comment:
We are too reliant on boolean flags, where actually what we want to represent is just the current state. Use an `enum`, with states like the following: `SCREENSAVER_DISABLED, SCREENSAVER_RUNNING, SCREENSAVER_STOPPING`

This way, we have fewer bools we need to keep track of.